### PR TITLE
Added DanswerBot response limit environment variables

### PR DESCRIPTION
**Pull request by hagen-danswer**
## Description
This adds a rate limit to the slack bot configurable by env variable

## How Has This Been Tested?
Turned the limit/time period down to 2 messages per 30 seconds and tried exceeding it.

## Accepted Risk
    High traffic at the end of one period and start of another could cause
    the limit to be exceeded.

### CODE REVIEW

File: backend/danswer/configs/danswerbot_configs.py [hunk 1]
```diff
@@ -73,3 +73,15 @@
 DANSWER_BOT_REPHRASE_MESSAGE = (
     os.environ.get("DANSWER_BOT_REPHRASE_MESSAGE", "").lower() == "true"
 )
+
```
Comment (hagen-danswer) [inline backend/danswer/configs/danswerbot_configs.py:76]: read comments

File: backend/danswer/danswerbot/slack/listener.py [hunk 2]
```diff
@@ -130,9 +131,19 @@ def prefilter_requests(req: SocketModeRequest, client: SocketModeClient) -> bool
 
         if event_type == "message":
             bot_tag_id = get_danswer_bot_app_id(client.web_client)
+
```
Comment (hagen-danswer) [inline backend/danswer/danswerbot/slack/listener.py:134]: this was just to make reading the logic a little easier

File: backend/danswer/danswerbot/slack/listener.py [hunk 3]
```diff
@@ -130,9 +131,19 @@ def prefilter_requests(req: SocketModeRequest, client: SocketModeClient) -> bool
...
+            is_dm = event.get("channel_type") == "im"
+            is_tagged = bot_tag_id and bot_tag_id in msg
+            is_danswer_bot_msg = bot_tag_id and bot_tag_id in event.get("user", "")
+
```
Comment (hagen-danswer) [inline backend/danswer/danswerbot/slack/listener.py:138]: read comment

File: backend/danswer/danswerbot/slack/utils.py [hunk 4]
```diff
@@ -41,7 +47,41 @@
 logger = setup_logger()
 
 
-DANSWER_BOT_APP_ID: str | None = None
+_DANSWER_BOT_APP_ID: str | None = None
+_DANSWER_BOT_MESSAGE_COUNT: int = 0
+_DANSWER_BOT_COUNT_START_TIME: float = time.time()
+
+
```
Comment (hagen-danswer) [inline backend/danswer/danswerbot/slack/utils.py:54]: just moved this function to the top near the global variables
renamed global variables for clarity

File: backend/danswer/danswerbot/slack/utils.py [hunk 5]
```diff
@@ -41,7 +47,41 @@
...
+        _DANSWER_BOT_APP_ID = web_client.auth_test().get("user_id")
+    return _DANSWER_BOT_APP_ID
+
+
+def check_message_limit() -> bool:
+    """
+    This isnt a perfect solution.
+    High traffic at the end of one period and start of another could cause
+    the limit to be exceeded.
+    """
```
Comment (hagen-danswer) [inline backend/danswer/danswerbot/slack/utils.py:67]: keeps count of messages sent
resets counter if the time period indicated has passed

**Review by hagen-danswer — COMMENTED** [review]
read comments

File: backend/danswer/configs/danswerbot_configs.py (hunk 1 above)
Comment (hagen-danswer) [reply backend/danswer/configs/danswerbot_configs.py:76]: currently defaults to 100,000 per 1 hour maximum

File: backend/danswer/danswerbot/slack/utils.py [hunk 6]
```diff
@@ -41,7 +47,41 @@
...
+
+
+def get_danswer_bot_app_id(web_client: WebClient) -> Any:
+    global _DANSWER_BOT_APP_ID
+    if _DANSWER_BOT_APP_ID is None:
+        _DANSWER_BOT_APP_ID = web_client.auth_test().get("user_id")
+    return _DANSWER_BOT_APP_ID
+
+
+def check_message_limit() -> bool:
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/utils.py:62]: would be nice to document what returning True or False means for the caller in a docstring.

File: backend/danswer/danswerbot/slack/utils.py [hunk 7]
```diff
@@ -41,7 +47,41 @@
 logger = setup_logger()
 
 
-DANSWER_BOT_APP_ID: str | None = None
+_DANSWER_BOT_APP_ID: str | None = None
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/utils.py:50]: Think we should be avoiding usage of globals here as this isn't going to generalize or be reusable easily.

File: backend/danswer/danswerbot/slack/utils.py [hunk 8]
```diff
@@ -41,7 +47,41 @@
...
+    This isnt a perfect solution.
+    High traffic at the end of one period and start of another could cause
+    the limit to be exceeded.
+    """
+    if DANSWER_BOT_RESPONSE_LIMIT_PER_TIME_PERIOD == 0:
+        return True
+    global _DANSWER_BOT_MESSAGE_COUNT
+    global _DANSWER_BOT_COUNT_START_TIME
+    time_since_start = time.time() - _DANSWER_BOT_COUNT_START_TIME
+    if time_since_start > DANSWER_BOT_RESPONSE_LIMIT_TIME_PERIOD_SECONDS:
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/utils.py:73]: I suspect we could import a less "chunky" rate limiting implementation from a third party.

**Review by rkuo-danswer — APPROVED** [review]
Comments inline ... in general, we should look into bringing in a reusable rate limiting implementation rather than rolling our own so maybe we don't address the individual comments.

For a quick fix we can put this out, but we should look into making this better soon. Maybe add a linear issue for cleanup next week?

**Review by Weves — APPROVED** [review]

**Pull request merged**